### PR TITLE
Terrain / Tile fixes

### PR DESCRIPTION
**Pull request by fmatthew5876**
* Add some new chipset BLOCK constants
* Add function for chip id -> chip index mapping
* If out of bounds coordinates are requested on non-looping map,
  RPG_RT will use the terrain of the first tile.
* Fix bug where tile substitution for last few tiles
  was broken for terrain lookups. Fix: #1834

### CODE REVIEW

**Comment by Ghabry** [comment]
Thanks for adding the constants and nice find with the "Tile 0" when out-of-bounds tiles are checked.

File: src/game_map.cpp [hunk 1]
```diff
@@ -847,6 +834,28 @@ int Game_Map::GetTerrainTag(int x, int y) {
...
+	// for out of bounds coordinates.
+	unsigned chip_index = 0;
+
+	if (Game_Map::IsValid(x, y)) {
+		const auto chip_id = map->lower_layer[x + y * GetWidth()];
+		chip_index = ChipIdToIndex(chip_id);
+
+		// Apply tile substitution
+		if (chip_index >= BLOCK_E_INDEX && chip_index < NUM_LOWER_TILES) {
+			chip_index = map_info.lower_tiles[chip_index - BLOCK_E_INDEX] + 18;
```
Comment (Ghabry) [inline src/game_map.cpp]: is this 18 something different to BLOCK_E_INDEX?

Comment (fmatthew5876) [reply src/game_map.cpp]: Nope, just an oversight. Fixed Thanks!

**Comment by Ghabry** [comment]
I ran a script to compare old and new, here are the values that mismatch @fmatthew5876 .

3150-3999 and 4600-4999 look harmless but for whole Block F the calculated chip ID looks wrong?

```
3150	6	0 <- BLOCK_C_END
-
3999	22	0 <- BLOCK_D - 1

4600	18	0 <- BLOCK_D_END
-
4999	25	0 <- BLOCK_E - 1

10000	0	162 <- BLOCK_F
-
10143	0	305 <- BLOCK_F_END
```

**Comment by Ghabry** [comment]
maybe the old code didn't calculate that chip ID at all? Is completely 0 in the whole range.

**Comment by fmatthew5876** [comment]
The old code did not calculate for block_f since it was inlined for lower chips only. I added this for completeness.

File: src/game_map.cpp [hunk 2]
```diff
@@ -847,6 +834,28 @@ int Game_Map::GetTerrainTag(int x, int y) {
 		return 1;
 	}
 
+	// Terrain tag wraps on looping maps
+	if (Game_Map::LoopHorizontal()) {
```
Comment (Ghabry) [inline src/game_map.cpp]: This LoopHorizontal/LoopVertical checks are unnecessary, RoundX and RoundY already do them.

**Comment by Ghabry** [comment]
Python code for completeness:

```py
BLOCK_A = 0;
BLOCK_A_TILES = 2;
BLOCK_A_STRIDE = 1000;
BLOCK_A_INDEX = 0;
BLOCK_A_END = BLOCK_A + BLOCK_A_TILES * BLOCK_A_STRIDE;

BLOCK_B = 2000;
BLOCK_B_TILES = 1;
BLOCK_B_STRIDE = 1000;
BLOCK_B_INDEX = 2;
BLOCK_B_END = BLOCK_B + BLOCK_B_TILES * BLOCK_B_STRIDE;

BLOCK_C = 3000;
BLOCK_C_TILES = 3;
BLOCK_C_STRIDE = 50;
BLOCK_C_INDEX = 3;
BLOCK_C_END = BLOCK_C + BLOCK_C_TILES * BLOCK_C_STRIDE;

BLOCK_D = 4000;
BLOCK_D_TILES = 12;
BLOCK_D_STRIDE = 50;
BLOCK_D_INDEX = 6;
BLOCK_D_END = BLOCK_D + BLOCK_D_TILES * BLOCK_D_STRIDE;

BLOCK_E = 5000;
BLOCK_E_TILES = 144;
BLOCK_E_STRIDE = 1;
BLOCK_E_INDEX = 18;
BLOCK_E_END = BLOCK_E + BLOCK_E_TILES * BLOCK_E_STRIDE;

BLOCK_F = 10000;
BLOCK_F_TILES = 144;
BLOCK_F_STRIDE = 1;
BLOCK_F_INDEX = 162;
BLOCK_F_END = BLOCK_F + BLOCK_F_TILES * BLOCK_F_STRIDE;

NUM_LOWER_TILES = BLOCK_F_INDEX;
NUM_UPPER_TILES = BLOCK_F_TILES;
NUM_TILES = NUM_LOWER_TILES + NUM_UPPER_TILES;


def chip_to_idx_old(chipID):
	if chipID < 3050:
		return 0 + chipID/1000
	elif chipID <  4000:
		return 4 + (chipID-3050)/50
	elif chipID <  5000:
		return 6 + (chipID-4000)/50
	elif chipID <  5144:
		return 18 + (chipID-5000)
	return 0;


def chip_to_idx_new(chip_id):
	if chip_id >= BLOCK_A and chip_id < BLOCK_A_END:
		return BLOCK_A_INDEX + (chip_id - BLOCK_A) / BLOCK_A_STRIDE;
	elif chip_id >= BLOCK_B and chip_id < BLOCK_B_END:
		return BLOCK_B_INDEX + (chip_id - BLOCK_B) / BLOCK_B_STRIDE;
	elif chip_id >= BLOCK_C and chip_id < BLOCK_C_END:
		return BLOCK_C_INDEX + (chip_id - BLOCK_C) / BLOCK_C_STRIDE;
	elif chip_id >= BLOCK_D and chip_id < BLOCK_D_END:
		return BLOCK_D_INDEX + (chip_id - BLOCK_D) / BLOCK_D_STRIDE;
	elif chip_id >= BLOCK_E and chip_id < BLOCK_E_END:
		return BLOCK_E_INDEX + (chip_id - BLOCK_E) / BLOCK_E_STRIDE;
	elif chip_id >= BLOCK_F and chip_id < BLOCK_F_END:
		return BLOCK_F_INDEX + (chip_id - BLOCK_F) / BLOCK_F_STRIDE;
	return 0


mm=[]
for i in range(100000):
  if int(chip_to_idx_old(i)) != int(chip_to_idx_new(i)):
    mm.append([i, int(chip_to_idx_old(i)), int(chip_to_idx_new(i))])
```

**Comment by fmatthew5876** [comment]
Your numbers and python code make sense to me. So I approve my own PR. 😀

**Pull request merged**